Enable capturing of comm collective parameters (#98)

louisfeng opened the pull request:
Summary:
X-link: https://github.com/facebookresearch/torch_ucc/pull/98

Add tensor input, output, and other metadata for PyTorch comms.

Test Plan: P517138779

Reviewed By: Pavani-Panakanti

Differential Revision: D38357077

H-Huang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -675,6 +677,17 @@ ProcessGroupNCCL::ProcessGroupNCCL(
             << "\nUSE_HIGH_PRIORITY_STREAM: "
             << options_->is_high_priority_stream;
 
+  RECORD_PARAM_COMMS(
```
Reason for adding this? (since initialization is not a collective)

louisfeng replied:
The network.ai team requested this feature. We can collect when a network collective is created during initialization. cc: Pavani.

H-Huang commented on torch/csrc/distributed/c10d/ParamCommsUtils.hpp:
```diff
@@ -65,17 +65,55 @@ class TORCH_API ParamCommsDebugInfo
-    outSplitSizes); \
+#define RECORD_PARAM_COMMS(                                                    \
+    seq,                                                                       \
+    WorkNccl,                                                                  \
```
Since this is a general comm util. It probably shouldn't be named WorkNCCL (a backend specific object) and make instead named to just `Work` instead? I have another comment to clarify what is actually getting logged here.

louisfeng replied:
Yes, makes sense, will fix.

louisfeng replied:
This supposed to be the ProcessGroup object, with the seq number can uniquely identify each collective operation.

H-Huang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1882,7 +1905,13 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::broadcast(
+  RECORD_PARAM_COMMS_DATA(
+      static_cast<int>(
+          this->getSequenceNumberForGroup() + 1), // seq + 1 to match collective
+      reinterpret_cast<std::intptr_t>(this), // work obj ptr
```
Confused about recording `this`. Isn't this a pointer to ProcessGroupNCCL rather than a work obj? Same applies to the other collectives

louisfeng replied:
Modified the comment, the code is correct, but the comment was wrong. Should be the process group object.

H-Huang reviewed (commented):
Hi, thanks for the PR. I am actually not familiar with the original use case for `RECORD_PARAM_COMMS`. Where does it get logged and is there a way to see the before / after of this change? Thanks!

louisfeng commented:
> 

These are opt-in performance and observability data logged by the PyTorch profiler. They are usually written to trace files for a few iterations.

H-Huang reviewed (approved):
LGTM, thanks for updating

louisfeng committed:
Enable capturing of comm collective parameters (#85368)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85368

X-link: https://github.com/facebookresearch/torch_ucc/pull/98

Add tensor input, output, and other metadata for PyTorch comms.

Test Plan:
```
buck build mode/opt-split-dwarf -c fbcode.nvcc_arch=a100 //hpc/models/ads:ads_10x_launcher
buck-out/gen/hpc/models/ads/ads_10x_launcher.par -- +launcher=local launcher.num_trainers=2 +data_loader=random
```

P517138779

Reviewed By: Pavani-Panakanti

Differential Revision: D38357077

fbshipit-source-id: c5394ff04ebe6af223c0e0a69a9cb558cc4c3a3f
